Remove bgp dot-notation handlers

chrisvanheuveln opened the pull request:
I discovered a few issues while rewriting the bgp global beaker test:
- Our BGP providers were normalizing the ASN to AS-PLAIN but the nxos cli doesn't support switching between AS-PLAIN and AS-DOT notations, even when the asn is equivalent; ie. once a tag syntax is used you must continue using it:

```
n7k8(config)# router bgp 1.4
n7k8(config-router)# router bgp 65540
bgp is already running; tag is 1.4.
```

Thus, I'm removing the asnum normalizer methods `process_asnum` and `dot_to_big`
- Fixed a day 1 problem with the yaml regexp which failed to match against AS-DOT notation
- Fixed event_history_detail. We use `auto_default: false` with this one because the N7k doesn't show a value with show run all, which means config_get returns [nil, nil].
- Misc minitest fixes. All of the bgp minitests are now passing.

chrisvanheuveln committed:
Remove bgp dot-notation handlers

I discovered a few issues while rewriting the bgp global beaker test:

* Our BGP providers were normalizing the ASN to AS-PLAIN but the nxos cli doesn't support switching between AS-PLAIN and AS-DOT notations, even when the asn is equivalent; ie. once a tag syntax is used you must continue using it:

```
n7k8(config)# router bgp 1.4
n7k8(config-router)# router bgp 65540
bgp is already running; tag is 1.4.
```

Thus, I'm removing the asnum normalizer methods `process_asnum` and `dot_to_big`

* Fixed a day 1 problem with the yaml regexp which failed to match against AS-DOT notation

* Fixed event_history_detail. We use `auto_default: false` with this one because the N7k doesn't show a value with show run all, which means config_get returns [nil, nil].

* Misc minitest fixes. All of the bgp minitests are now passing.

robert-w-gries commented on lib/cisco_node_utils/bgp.rb:
```diff
@@ -26,7 +26,7 @@ class RouterBgp < NodeUtil
       fail ArgumentError unless vrf.is_a? String
       fail ArgumentError unless vrf.length > 0
-      @asnum = RouterBgp.process_asnum(asnum)
+      @asnum = asnum.to_s
```
Seeing this error on 3k now

```
Node under test:
  - name  - n3k-51
  - type  - N3K-C3048TP-1GE
  - image - 

TestRouterBgp#test_routerbgp_get_asnum = 10.40 s = F

Finished in 10.406674s, 0.0961 runs/s, 0.0961 assertions/s.

  1) Failure:
TestRouterBgp#test_routerbgp_get_asnum [tests/test_router_bgp.rb:154]:
Error: router asnum not correct.
Expected: 55
  Actual: "55"
```

mikewiebe replied:
Looks good other then this.

chrisvanheuveln replied:
fixed

glennmatthews commented:
XR CLI has the same restriction, it appears, so I'm in favor of this change:

```
router bgp 65540
!!% The instance name is used already: asn 1.4 inst-name default
```

glennmatthews commented:
I'm a bit surprised that this scenario wasn't covered by the testing before. Since you're removing the old functionality, does it make sense to add some negative tests around trying to change the format?

Chris: I'll add a test or two.

glennmatthews commented on lib/cisco_node_utils/bgp.rb:
```diff
@@ -36,19 +36,6 @@ def initialize(asnum, vrf='default', instantiate=true)
-                                         asnum.is_a?(String)
-      if asnum.is_a? String
-        # Match ASDOT '1.5' or ASPLAIN '55' strings
-        fail ArgumentError unless /^(\d+|\d+\.\d+)$/.match(asnum)
```
I see that you've changed the tests to look for a CliError instead of this ArgumentError but IMHO the ArgumentError is more intuitive to troubleshoot. Would it make sense to keep the validation logic (but not the conversion) in this function and rename it to `self.validate_asnum`?

chrisvanheuveln replied:
Sure that sounds reasonable. Of course now I get to put all those calls back in :-)

chrisvanheuveln committed:
Add BGP validate_asnum, fix tests

Oddly, the maxas test failed with my changes until I noticed that it was using different asnum vals for the default and a vrf. I'm not sure how that could have worked before. It's fixed now.

mikewiebe commented:
:+1:
